src/main/java/picard/sam/SamAlignmentMerger.java

```diff
@@ -158,6 +141,70 @@ public void mergeAlignment(final File referenceFasta) {
         }
     }
 
+    @Override
+    protected SAMSequenceDictionary getDictionaryForMergedBam() {
+        SAMSequenceDictionary finalDict = null;
+        if (alignedSamFile != null && !alignedSamFile.isEmpty()) {
+            for (final File alignedFile : alignedSamFile) {
+                final SAMSequenceDictionary dict = SAMSequenceDictionaryExtractor.extractDictionary(alignedFile);
+                if (finalDict == null) {
+                    finalDict = dict;
+                } else {
+                    if (!compareSequenceDictionaries(finalDict, dict)) {
+                        log.error("Dictionaries of input files differ. Unable to continue.");
+                        throw new UnsupportedOperationException("Cannot use input files with different dictionaries.");
+                    }
+                }
+            }
+        } else {
+            final SeparateEndAlignmentIterator mergingIterator = new SeparateEndAlignmentIterator(this.read1AlignedSamFile, this.read2AlignedSamFile, referenceFasta);
+            finalDict = mergingIterator.getHeader().getSequenceDictionary();
+        }
+
+        SAMSequenceDictionary referenceDict = SAMSequenceDictionaryExtractor.extractDictionary(referenceFasta);
+        if (referenceDict == null) {
+            throw new PicardException("No sequence dictionary found for " + referenceFasta.getAbsolutePath() +
+                    ".  Use CreateSequenceDictionary.jar to create a sequence dictionary.");
+        }
+        return mergeDictionaries(finalDict, SAMSequenceDictionaryExtractor.extractDictionary(referenceFasta));
+    }
+
+    private SAMSequenceDictionary mergeDictionaries(SAMSequenceDictionary dict1, SAMSequenceDictionary dict2) {
+
+        final SAMSequenceDictionary finalDict = new SAMSequenceDictionary(dict1.getSequences());
+        for (final SAMSequenceRecord sequence : finalDict.getSequences()) {
+            final int sequenceIndex = sequence.getSequenceIndex();
+            final Set<String> allTags = new HashSet<>();
+
+            for (SAMSequenceRecord seq : Arrays.asList(
+                    dict1.getSequence(sequenceIndex),
+                    dict2.getSequence(sequenceIndex))) {
+
+                final Set<String> dictTags = seq
+                        .getAttributes().parallelStream()
+                        .map(Map.Entry::getKey)
+                        .collect(Collectors.toSet());
+                allTags.addAll(dictTags);
+            }
+
+            for (final String tag : allTags) {
+                final String value1 = dict1.getSequence(sequenceIndex).getAttribute(tag);
+                final String value2 = dict2.getSequence(sequenceIndex).getAttribute(tag);
+
+                if (value1 != null && value2 != null && !value1.equals(value2)) {
```

Comment:
I'd be more inclined to go the opposite way and assert that the name, LN and MD tags that really should match, and just warn about mismatches on other attributes.